lib/utils/char-set-tree.ts

```diff
@@ -0,0 +1,357 @@
+                    state = {
+                        addNext: (t) =>
+                            groupList.forEach((tree) =>
+                                tree.addNextInternal(t),
+                            ),
+                        markAsEndOfPattern: () =>
+                            groupList.forEach((tree) =>
+                                tree.markAsEndOfPatternInternal(),
+                            ),
+                    }
+                } else if (element.type === "Quantifier") {
+                    if (element.max === 0) {
+                        continue
+                    }
+                    if (element.max > 10) {
+                        // consider has complex pattern
+                        return null
+                    }
+                    const elementList = elementsToList([element.element])
+                    if (!elementList) {
+                        // has complex pattern
+                        return null
+                    }
+                    for (const tree of elementList.list) {
+                        state.addNext(tree)
+                        tree.setRangeInternal(element.min, element.max)
+                    }
+
+                    const newState: State = {
+                        addNext: (t) =>
+                            elementList.list.forEach((tree) =>
+                                tree.addNextInternal(t),
+                            ),
+                        markAsEndOfPattern: () =>
+                            elementList.list.forEach((tree) =>
+                                tree.markAsEndOfPatternInternal(),
+                            ),
+                    }
+
+                    if (element.min === 0) {
+                        const selfState = state
+                        state = {
+                            addNext(t) {
+                                selfState.addNext(t)
+                                newState.addNext(t)
+                            },
+                            markAsEndOfPattern() {
+                                selfState.markAsEndOfPattern()
+                                newState.markAsEndOfPattern()
+                            },
+                        }
+                    } else {
+                        state = newState
+                    }
+                } else {
+                    // has complex pattern
+                    return null
+                }
+            }
+
+            return {
+                list,
+                markAsEndOfPattern: state.markAsEndOfPattern,
+            }
+        }
+
+        /** Element to tree */
+        function toTree(element: ToCharSetElement): CharSetTree {
+            return new CharSetTree({
+                element,
+                charSet: toCharSet(element, regexpContext.flags),
+            })
+        }
+    }
+
+    /**
+     * Checks whether the given tree is included.
+     */
+    public includes(target: CharSetTreeRoot): boolean {
```

Comment:
Please explain this operation to me. I get how this works for simple cases but I don't see how it generalizes to alternations and quantifiers. What does it mean for one tree to _include_ another?